lib/internal/modules/cjs/loader.js

```diff
@@ -960,13 +960,24 @@ Module.prototype._compile = function(content, filename) {
   return result;
 };
 
-
 // Native extension for .js
+let warnRequireESM = true;
 Module._extensions['.js'] = function(module, filename) {
-  if (experimentalModules && filename.endsWith('.js')) {
+  if (filename.endsWith('.js')) {
     const pkg = readPackageScope(filename);
     if (pkg && pkg.type === 'module') {
-      throw new ERR_REQUIRE_ESM(filename);
+      if (warnRequireESM) {
+        process.emitWarning(
+          'require() of .js files inside of a package with "type": "module" ' +
+          'in its package.json is not supported under --experimental-modules' +
+          '\nRather use import to load this module, or if you are the author' +
+          ' you may want to use the .cjs extension for this file.'
```

Comment:
We would need to rewrite this warning when `--experimental-modules` is unflagged. Could we perhaps come up with language that can ship now and won’t need to be changed later?

I’m also thinking that it might be good to include the path/filename of the .js file being `require`d, and the path/filename of the `require`r, for example `./.eslintrc.js` required by `./node_modules/eslint/bin/eslint.js` or whatever. The latter is already in the stack trace, but including it in the message might make the connection more obvious.

So how about something like:

```
<requiring path/file> contains a require() call of the .js file <required path/file>.
require() of a .js file whose nearest parent package.json contains "type": "module"
is deprecated and will error in the future. Instead, rename the required file to end
in .cjs; change the requiring code to use import(); or remove "type": "module" from
the required file's nearest parent package.json.
```

Reply:
I've added the message context here, let me know if that works for you.